src/goto-symex/goto_symex.cpp

```diff
@@ -87,8 +96,277 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
     if(state.source.pc->source_location.get_hide())
       assignment_type = symex_targett::assignment_typet::HIDDEN;
 
+    symex_assignt symex_assign{
+      state, assignment_type, ns, symex_config, target};
+
+    // Try to constant propagate potential side effects of the assignment, when
+    // simplification is turned on and there is one thread only. Constant
+    // propagation is only sound for sequential code as here we do not take into
+    // account potential writes from other threads when propagating the values.
+    if(symex_config.simplify_opt && state.threads.size() <= 1)
+    {
+      if(constant_propagate_assignment_with_side_effects(
+           state, symex_assign, lhs, rhs))
+        return;
+    }
+
     exprt::operandst lhs_if_then_else_conditions;
-    symex_assignt{state, assignment_type, ns, symex_config, target}.assign_rec(
+    symex_assign.assign_rec(
       lhs, expr_skeletont{}, rhs, lhs_if_then_else_conditions);
   }
 }
+
+bool goto_symext::constant_propagate_assignment_with_side_effects(
+  statet &state,
+  symex_assignt &symex_assign,
+  const exprt &lhs,
+  const exprt &rhs)
+{
+  if(rhs.id() == ID_function_application)
+  {
+    exprt rhs_l2 = state.rename(rhs, ns).get();
+
+    const function_application_exprt &f_l2 =
+      to_function_application_expr(rhs_l2);
+
+    const function_application_exprt &f_l1 = to_function_application_expr(rhs);
+
+    if(f_l1.function().id() == ID_symbol)
+    {
+      const irep_idt &func_id = f_l1.function().get_identifier();
+
+      if(func_id == ID_cprover_string_concat_func)
+      {
+        return constant_propagate_string_concat(
+          state, symex_assign, f_l1, f_l2);
+      }
+      else if(func_id == ID_cprover_string_empty_string_func)
+      {
+        // constant propagating the empty string always succeeds as it does
+        // not depend on potentially non-constant inputs
+        constant_propagate_empty_string(state, symex_assign, f_l1);
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+void goto_symext::assign_string_constant(
+  statet &state,
+  symex_assignt &symex_assign,
+  const ssa_exprt &length,
+  const constant_exprt &new_length,
+  const ssa_exprt &char_array,
+  const array_exprt &new_char_array)
+{
```

Comment:
:pick: It looks like this function could do with some refactoring. It's a bit too long to be readable. Extracting parts into named functions would make it more understandable.

Reply:
Ok, added to helper functions being called by this function.